src/tag-input/src/TagInput.js

```diff
@@ -166,30 +176,118 @@ const TagInput = memo(
         ref={ref}
         onBlur={handleBlur}
         {...boxProps}
         {...rest}
+        paddingRight={hasAutocomplete ? majorScale(3) : undefined}
       >
         {values.map(maybeRenderTag)}
-        <TextInput
-          appearance="none"
-          id={id}
-          disabled={disabled}
-          flexGrow="1"
-          height={height - 4}
-          width="auto"
-          type="text"
-          value={inputValue}
-          {...inputProps}
-          ref={inputRef}
-          onChange={handleInputChange}
-          onFocus={handleInputFocus}
-          onKeyDown={handleKeyDown}
-        />
+        <Box flexGrow="1" display="inline-block">
+          <Autocomplete
+            onChange={changedItem => {
+              addTags(changedItem)
+              setInputValue('')
+            }}
+            items={hasAutocomplete ? autocompleteItems : []}
+            id={inputId}
+            selectedItem=""
+            inputValue={inputValue}
+          >
+            {autocompleteProps => {
+              const {
+                closeMenu,
+                getInputProps,
+                getRef: autocompleteGetRef,
+                getToggleButtonProps,
+                highlightedIndex
+              } = autocompleteProps
+
+              const {
+                onBlur: autocompleteOnBlur,
+                onChange: autocompleteOnChange,
+                onKeyDown: autocompleteKeyDown,
+                ...autocompleteRestProps
+              } = getInputProps()
+
+              return (
+                <>
+                  <TextInput
+                    appearance="none"
+                    disabled={disabled}
+                    height={height - 4}
+                    width="auto"
+                    type="text"
+                    {...inputProps}
+                    {...autocompleteRestProps}
+                    value={inputValue}
+                    id={inputId}
+                    ref={textInputRef => {
+                      autocompleteGetRef(textInputRef)
+                      if (inputRef instanceof Function) {
+                        inputRef(textInputRef)
+                      } else if (inputRef) {
+                        inputRef.current = textInputRef
+                      }
+                    }}
+                    onBlur={e => {
+                      autocompleteOnBlur(e)
+                      if (inputProps.onBlur) {
+                        safeInvoke(inputProps.onBlur, e)
+                      }
+                    }}
+                    onFocus={e => {
+                      handleInputFocus(e)
+                      if (inputProps.onFocus) {
+                        safeInvoke(inputProps.onFocus, e)
+                      }
+                    }}
+                    onChange={e => {
+                      handleInputChange(e)
+                      autocompleteOnChange(e)
+                    }}
+                    onKeyDown={e => {
```

Comment:
Could this handler be pulled up higher to reduce nesting and improve readability? Not sure what all it needs in this scope, but even pulling it up into the child render of `Autocomplete` (lines 201-215ish) would help, imo.

Reply:
Sure, that makes sense. It does need the `closeMenu` function, so I'll keep it within the `Autocomplete`.